IncrementalSweeper should zero memory when it sweeps MarkedBlocks.

kmiller68 opened the pull request:
#### 5b799ef3e5c7005e84d68250551b430b11c43c58
<pre>
IncrementalSweeper should zero memory when it sweeps MarkedBlocks.
<a href="https://bugs.webkit.org/show_bug.cgi?id=279063">https://bugs.webkit.org/show_bug.cgi?id=279063</a>
<a href="https://rdar.apple.com/135198347">rdar://135198347</a>

Reviewed by NOBODY (OOPS!).

This patch adds a new SweepMode, SweepAndZero, which causes our
sweeper to zero the memory it swept. This should help the memory
compressor on platforms that have one do a better job compressing
memory. The only places where we use this new mode are for
IncrementalSweeper and Synchronous GC. The latter is historically
mostly used when we really want to reduce memory pressure so zeroing
will likely be profitable.

This patch also makes the template specialization for specializedSweep
a struct, which I think makes the code a bit easier to follow and requires
less code duplication when adding new modes.

* Source/JavaScriptCore/heap/BlockDirectory.cpp:
(JSC::BlockDirectory::sweep):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/HeapCellType.cpp:
(JSC::HeapCellType::finishSweep const):
* Source/JavaScriptCore/heap/HeapCellType.h:
* Source/JavaScriptCore/heap/IncrementalSweeper.cpp:
(JSC::IncrementalSweeper::sweepNextBlock):
* Source/JavaScriptCore/heap/IsoHeapCellType.cpp:
(JSC::IsoHeapCellType::finishSweep const):
* Source/JavaScriptCore/heap/IsoHeapCellType.h:
* Source/JavaScriptCore/heap/IsoInlinedHeapCellType.h:
* Source/JavaScriptCore/heap/IsoInlinedHeapCellTypeInlines.h:
(JSC::IsoInlinedHeapCellType&lt;CellType&gt;::finishSweep const):
* Source/JavaScriptCore/heap/LocalAllocator.cpp:
(JSC::LocalAllocator::tryAllocateWithoutCollecting):
(JSC::LocalAllocator::tryAllocateIn):
* Source/JavaScriptCore/heap/MarkedBlock.cpp:
(JSC::MarkedBlock::Handle::lastChanceToFinalize):
(JSC::MarkedBlock::Handle::resumeAllocating):
(JSC::MarkedBlock::Handle::sweep):
(WTF::printInternal): Deleted.
* Source/JavaScriptCore/heap/MarkedBlock.h:
* Source/JavaScriptCore/heap/MarkedBlockInlines.h:
(JSC::MarkedBlock::Handle::SpecializedSweepConfig::with const):
(JSC::MarkedBlock::Handle::specializedSweep):
(JSC::MarkedBlock::Handle::finishSweepKnowingHeapCellType):
* Source/JavaScriptCore/heap/MarkedSpace.cpp:
(JSC::MarkedSpace::lastChanceToFinalize):
* Source/JavaScriptCore/heap/Subspace.cpp:
(JSC::Subspace::finishSweep): Deleted.
* Source/JavaScriptCore/heap/Subspace.h:
* Source/JavaScriptCore/runtime/JSDestructibleObjectHeapCellType.cpp:
(JSC::JSDestructibleObjectHeapCellType::finishSweep const):
* Source/JavaScriptCore/runtime/JSDestructibleObjectHeapCellType.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5b799ef3e5c7005e84d68250551b430b11c43c58

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65479 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44850 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18097 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69505 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16088 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52651 "Built successfully") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16368 "Hash 5b799ef3 for PR 33069 does not build (failure)") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52587 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11160 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68546 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41442 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56676 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33213 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38117 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14051 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14964 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/58593 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59952 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14392 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71210 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/64723 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9433 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/61/builds/16368 "Hash 5b799ef3 for PR 33069 does not build (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59905 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9465 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56737 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60177 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/uri-scheme (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7796 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1448 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/86490 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40660 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15233 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41736 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42919 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41480 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (commented):
Can you show me the performance number on this change in SP3 / JS2 / PLUM3 / Membuster5?

kmiller68 committed:
IncrementalSweeper should zero memory when it sweeps MarkedBlocks.
https://bugs.webkit.org/show_bug.cgi?id=279063
rdar://135198347

Reviewed by NOBODY (OOPS!).

This patch adds a new SweepMode, SweepAndZero, which causes our
sweeper to zero the memory it swept. This should help the memory
compressor on platforms that have one do a better job compressing
memory. The only places where we use this new mode are for
IncrementalSweeper and Synchronous GC. The latter is historically
mostly used when we really want to reduce memory pressure so zeroing
will likely be profitable.

This patch also makes the template specialization for specializedSweep
a struct, which I think makes the code a bit easier to follow and requires
less code duplication when adding new modes.

* Source/JavaScriptCore/heap/BlockDirectory.cpp:
(JSC::BlockDirectory::sweep):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/HeapCellType.cpp:
(JSC::HeapCellType::finishSweep const):
* Source/JavaScriptCore/heap/HeapCellType.h:
* Source/JavaScriptCore/heap/IncrementalSweeper.cpp:
(JSC::IncrementalSweeper::sweepNextBlock):
* Source/JavaScriptCore/heap/IsoHeapCellType.cpp:
(JSC::IsoHeapCellType::finishSweep const):
* Source/JavaScriptCore/heap/IsoHeapCellType.h:
* Source/JavaScriptCore/heap/IsoInlinedHeapCellType.h:
* Source/JavaScriptCore/heap/IsoInlinedHeapCellTypeInlines.h:
(JSC::IsoInlinedHeapCellType<CellType>::finishSweep const):
* Source/JavaScriptCore/heap/LocalAllocator.cpp:
(JSC::LocalAllocator::tryAllocateWithoutCollecting):
(JSC::LocalAllocator::tryAllocateIn):
* Source/JavaScriptCore/heap/MarkedBlock.cpp:
(JSC::MarkedBlock::Handle::lastChanceToFinalize):
(JSC::MarkedBlock::Handle::resumeAllocating):
(JSC::MarkedBlock::Handle::sweep):
(WTF::printInternal): Deleted.
* Source/JavaScriptCore/heap/MarkedBlock.h:
* Source/JavaScriptCore/heap/MarkedBlockInlines.h:
(JSC::MarkedBlock::Handle::SpecializedSweepConfig::with const):
(JSC::MarkedBlock::Handle::specializedSweep):
(JSC::MarkedBlock::Handle::finishSweepKnowingHeapCellType):
* Source/JavaScriptCore/heap/MarkedSpace.cpp:
(JSC::MarkedSpace::lastChanceToFinalize):
* Source/JavaScriptCore/heap/Subspace.cpp:
(JSC::Subspace::finishSweep): Deleted.
* Source/JavaScriptCore/heap/Subspace.h:
* Source/JavaScriptCore/runtime/JSDestructibleObjectHeapCellType.cpp:
(JSC::JSDestructibleObjectHeapCellType::finishSweep const):
* Source/JavaScriptCore/runtime/JSDestructibleObjectHeapCellType.h:
